src/block_ffm.rs

```diff
@@ -879,49 +1074,91 @@ impl<L: OptimizerTrait + 'static> BlockFFM<L> {
 
-            *ffm_slice.get_unchecked_mut(self_interaction_index + f1) += v * 0.5;
+                contra_field = hadd_ps(_mm_add_ps(acc_0, acc_1));
+            } else {
+                for _ in (0..ffmk_end_as_usize).step_by(LANES) {
+                    let contra_field_0 = _mm_loadu_ps(contra_fields_ptr);
+                    contra_fields_ptr = contra_fields_ptr.add(STEP);
+                    let contra_field_1 = _mm_loadu_ps(contra_fields_ptr);
+                    contra_fields_ptr = contra_fields_ptr.add(STEP);
 
-            let mut f2_index_offset = f1_index_offset + ffm_fields_count_as_usize;
-            let mut f2_offset_ffmk = f1_offset + f1_ffmk;
-            for f2 in f1 + 1..ffm_fields_count_as_usize {
-                let f1_index = f1_index_offset + f2;
-                let f2_index = f2_index_offset + f1;
+                    let acc_0 = _mm_mul_ps(contra_field_0, contra_field_0);
+                    let acc_1 = _mm_mul_ps(contra_field_1, contra_field_1);
 
-                f1_offset_ffmk += ffmk_as_usize;
-                f2_offset_ffmk += field_embedding_len_as_usize;
+                    contra_field += hadd_ps(_mm_add_ps(acc_0, acc_1));
+                }
 
-                let mut contra_field = 0.0;
-                for k in 0..ffmk_as_usize {
+                for k in ffmk_end_as_usize..ffmk_as_usize {
                     contra_field += contra_fields.get_unchecked(f1_offset_ffmk + k)
-                        * contra_fields.get_unchecked(f2_offset_ffmk + k);
+                        * contra_fields.get_unchecked(f1_offset_ffmk + k);
                 }
+            }
+            *ffm_slice.get_unchecked_mut(f1 * ffm_fields_count_as_usize + f1) += contra_field * 0.5;
 
-                contra_field = contra_field * 0.5;
+            let mut f2_offset_ffmk = f1_offset + f1_ffmk;
+            for f2 in f1 + 1..ffm_fields_count_as_usize {
+                f2_offset_ffmk += field_embedding_len_as_usize;
+                f1_offset_ffmk += ffmk_as_usize;
+
+                let mut contra_field = 0.0;
+                let mut contra_fields_ptr_1 = contra_fields.as_ptr().add(f1_offset_ffmk);
+                let mut contra_fields_ptr_2 = contra_fields.as_ptr().add(f2_offset_ffmk);
+                if ffmk_as_usize == LANES {
+                    let contra_field_0 = _mm_loadu_ps(contra_fields_ptr_1);
+                    let contra_field_1 = _mm_loadu_ps(contra_fields_ptr_2);
+                    let acc_0 = _mm_mul_ps(contra_field_0, contra_field_1);
+
+                    let contra_field_2 = _mm_loadu_ps(contra_fields_ptr_1.add(STEP));
+                    let contra_field_3 = _mm_loadu_ps(contra_fields_ptr_2.add(STEP));
+                    let acc_1 = _mm_mul_ps(contra_field_2, contra_field_3);
+
+                    contra_field = hadd_ps(_mm_add_ps(acc_0, acc_1));
+                } else {
+                    for _ in (0..ffmk_end_as_usize).step_by(LANES) {
```

Comment:
If I understand this right, this is not 1:1 op map to previous version of loop structure right? These intermediary additions + the final normalization most likely result in different rig?

Reply:
Actually it is the same. I only divided it more per use case where we have 8 & more (for future).
Created a new model with this binary based on another one & the rig was the same.

Reply:
Interesting, thanks for checking too!